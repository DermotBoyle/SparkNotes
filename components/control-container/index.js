export { ControlContainer } from './control-container.jsx'

export { SideMenu } from './side-menu.jsx'

export { ControlMenu } from './control-menu.jsx'
