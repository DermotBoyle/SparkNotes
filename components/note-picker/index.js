export { NotePicker } from './note-picker'
