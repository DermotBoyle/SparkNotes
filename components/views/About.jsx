import React from 'react'
import { useEffect, useState } from 'react'

export const About = ({ transitionStyles, defaultStyle, transitionState }) => {
  const [on, setOn] = useState('entering')

  useEffect(() => {
    setOn(transitionState)
  }, [])

  return (
    <section
      style={{
        ...defaultStyle,
        ...transitionStyles[on]
      }}
    >
      <p>This is an about App I built with NextJS and MongoDB</p>
    </section>
  )
}
