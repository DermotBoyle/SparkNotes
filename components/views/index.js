export { About } from './About'

export { Find } from './Find'

export { ViewAll } from './ViewAll'

export { Home } from './Home'

export { CreateNoteForm } from './CreateNoteForm'

export { EditNoteForm } from './EditNoteForm'
