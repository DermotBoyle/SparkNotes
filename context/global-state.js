import { createContext, useContext, useReducer } from 'react'
import { urlReducer } from './app-reducer'

const AppContext = createContext()

export function AppWrapper ({ children }) {
  const [state, dispatch] = useReducer(urlReducer, {
    currentURL: '/home'
  })

  const value = { state, dispatch }

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>
}

export function useAppContext () {
  return useContext(AppContext)
}
