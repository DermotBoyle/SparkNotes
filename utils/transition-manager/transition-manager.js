import React from 'react'
import {
  Home,
  ViewAll,
  About,
  Find,
  CreateNoteForm,
  EditNoteForm
} from 'components/views'
import { Routes } from 'utils'

export const transitionManager = transitionState => {
  const View = {
    [Routes.HOME]: (
      <Home
        transitionState={transitionState}
        defaultStyle={defaultStyle}
        transitionStyles={transitionStyles}
      />
    ),
    [Routes.VIEW_ALL]: (
      <ViewAll
        transitionState={transitionState}
        defaultStyle={defaultStyle}
        transitionStyles={transitionStyles}
      />
    ),
    [Routes.FIND]: (
      <Find
        transitionState={transitionState}
        defaultStyle={defaultStyle}
        transitionStyles={transitionStyles}
      />
    ),
    [Routes.ABOUT]: (
      <About
        transitionState={transitionState}
        defaultStyle={defaultStyle}
        transitionStyles={transitionStyles}
      />
    ),
    [Routes.CREATE]: (
      <CreateNoteForm
        transitionState={transitionState}
        defaultStyle={defaultStyle}
        transitionStyles={transitionStyles}
      />
    ),
    [Routes.EDIT]: (
      <EditNoteForm
        transitionState={transitionState}
        defaultStyle={defaultStyle}
        transitionStyles={transitionStyles}
      />
    )
  }

  return function ComponentToRender ({ currentURL }) {
    return View[currentURL]
  }
}

export const defaultStyle = {
  transition: `opacity 800ms ease-out`,
  opacity: 0
}

export const transitionStyles = {
  entered: { opacity: 1 },
  exiting: { opacity: 0 }
}
