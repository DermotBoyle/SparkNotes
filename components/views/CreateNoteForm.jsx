import { SaveButton } from 'components/save-button'
import { useReducer } from 'react'
import router from 'next/router'

import styles from 'styles/create-note-form.module.scss'

import moment from 'moment'

import { Routes } from 'utils'

export const CreateNoteForm = ({
  transitionStyles,
  defaultStyle,
  transitionState
}) => {
  const initialFormValues = {
    subject: '',
    keywords: '',
    content: '',
    date: ''
  }

  const [formValues, setFormValues] = useReducer(
    (curVals, newVals) => ({ ...curVals, ...newVals }),
    initialFormValues
  )

  const { subject, keywords, content } = formValues

  const handleFormChange = ({ target }) => {
    const { name, value } = target
    setFormValues({ [name]: value })
  }

  const handleSaveClick = async () => {
    formValues.date = moment(new Date()).format('YYYY-MM-DD-T00:00:00.000Z')
    updateDB(formValues)
  }

  const updateDB = async formValues => {
    try {
      await fetch('api/note', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(formValues)
      }).then(({ statusText }) => {
        if (statusText === 'Created') {
          setFormValues(initialFormValues)
          if (res.status === 201) router.push(Routes.BASE + Routes.VIEW_ALL)
        }
      })
    } catch (error) {
      console.error(error)
    }
  }

  return (
    <section
      className={styles['form-container']}
      style={{
        ...defaultStyle,
        ...transitionStyles[transitionState]
      }}
    >
      <form>
        <div className={styles['keywords-container']}>
          <div>
            <label>Subject</label>
            <input
              type='text'
              name='subject'
              value={subject}
              onChange={handleFormChange}
            />
          </div>
          <div>
            <label>Keywords</label>
            <input
              type='text'
              name='keywords'
              value={keywords}
              onChange={handleFormChange}
            />
          </div>
        </div>
        <div className={styles['main-note']}>
          <label>Write your note here</label>
          <textarea
            type='text'
            name='content'
            value={content}
            onChange={handleFormChange}
          />
        </div>
      </form>
      <SaveButton formValues={formValues} handleSaveClick={handleSaveClick} />
    </section>
  )
}
