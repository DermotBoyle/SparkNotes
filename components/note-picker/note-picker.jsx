import React from 'react'
import Head from 'next/head'
import Image from 'next/image'
import { useAppContext } from 'context/global-state'
import edit from 'public/edit-icon.svg'
import { useRouter } from 'next/router'

import { Routes } from 'utils'

import styles from 'styles/carousel-container.module.scss'

export const NotePicker = ({ allNotes, isLoading }) => {
  const { dispatch } = useAppContext()
  const router = useRouter()

  const goToEdit = note => {
    dispatch({
      type: 'update',
      payload: { currentNote: note, currentURL: 'edit-note' }
    })
    router.push(`${Routes.BASE}${Routes.EDIT}`)
  }

  const deleteNote = _id => {
    fetch('/api/note/' + _id, {
      method: 'DELETE'
    }).then(res => {
      res.status === 201 && router.reload()
    })
  }

  return (
    <>
      <Head>
        <link
          href='https://fonts.googleapis.com/css2?family=Open+Sans&family=Roboto&family=Satisfy&display=swap'
          rel='stylesheet'
        />
      </Head>
      <div className={styles.carousel}>
        {allNotes?.data?.length ? (
          allNotes?.data?.map(({ subject, content, _id, keywords = [] }) => (
            <div key={_id} className={styles.item}>
              <Image
                data-tip
                data-for='edit-button'
                src={edit}
                width={15}
                onClick={() => goToEdit({ subject, content, _id, keywords })}
              />
              <h5>{subject}</h5>
              <p style={{ fontFamily: 'Satisfy' }}>{content}</p>
              <button onClick={() => deleteNote({ _id })}>DELETE</button>
            </div>
          ))
        ) : !!isLoading ? (
          <Image src='/loader.svg' width={50} height={50} />
        ) : (
          <h3>No notes to show</h3>
        )}
      </div>
    </>
  )
}
