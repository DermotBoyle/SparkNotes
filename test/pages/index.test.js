/* eslint-disable react/display-name */
import React from 'react'
import { ViewAll, About, Find } from 'components/views'
import { transitionStyles } from 'utils/transition-manager/transition-manager'
import { fireEvent, render, act, screen } from '@testing-library/react'
import MainComponent from 'pages/home/index.js'
import { AppContext } from 'context/global-state'
import preloadAll from 'jest-next-dynamic'
import * as nextRouter from 'next/router'
import { NotePicker } from 'components/note-picker'

// eslint-disable-next-line react/display-name
// eslint-disable-next-line @next/next/no-img-element
jest.mock('next/image', () => ({ src, alt }) => <img src={src} alt={alt} />)
jest.mock('crypto', () => ({
  randomBytes: num => new Array(num).fill(0)
}))

jest.mock('utils/helpers/fetcher', () => ({
  fetcher: jest.fn().mockResolvedValueOnce([]),
  readData: jest.fn().mockResolvedValue([])
}))

const allNotes = {
  data: [
    {
      _id: 1,
      subject: 'testSubject',
      content: 'content',
      keywords: 'keywords',
      created: 'fakeDate',
      updated: 'fakeUpdated'
    },
    {
      _id: 2,
      subject: 'testSubject',
      content: 'content',
      keywords: 'keywords',
      created: 'fakeDate',
      updated: 'fakeUpdated'
    },
    {
      _id: 3,
      subject: 'testSubject',
      content: 'content',
      keywords: 'keywords',
      created: 'fakeDate',
      updated: 'fakeUpdated'
    }
  ]
}

const WrapperProviderMain = (mockDispatch, mockState) => {
  return (
    <AppContext.Provider
      value={{
        dispatch: mockDispatch,
        state: mockState
      }}
    >
      <MainComponent />
    </AppContext.Provider>
  )
}

const WrapperProviderComp = (mockDispatch, mockState) => {
  function WrappedComponent (component) {
    return (
      <AppContext.Provider
        value={{
          dispatch: mockDispatch,
          state: mockState
        }}
      >
        {component}
      </AppContext.Provider>
    )
  }
  return WrappedComponent
}

describe('<MainComponent/>', () => {
  beforeAll(async () => {
    await preloadAll()
  })

  it('should render the side menu', async () => {
    const mockDispatch = jest.fn()
    const mockState = { currentURL: '/home' }
    nextRouter.useRouter = jest.fn()
    nextRouter.useRouter.mockImplementation(() => ({
      route: '/home',
      query: { section: '/' }
    }))

    const { baseElement } = render(WrapperProviderMain(mockDispatch, mockState))

    const element = screen.getByTestId('side-menu')
    const HomeElement = screen.getByTestId('home')

    expect(HomeElement).toHaveClass('nav-link__active')
    expect(element).toBeInTheDocument()
    expect(baseElement).toMatchSnapshot()
  })

  it('should change view onClick of nav item', async () => {
    const mockDispatch = jest.fn()
    const mockState = { currentURL: '/home' }
    nextRouter.useRouter = jest.fn()
    nextRouter.useRouter.mockImplementation(() => ({
      route: '/home',
      query: { section: '/' },
      push: jest.fn()
    }))

    const { baseElement } = render(WrapperProviderMain(mockDispatch, mockState))

    const element = screen.getByText('view all')
    fireEvent.click(element)

    expect(element).toBeInTheDocument()
    expect(baseElement).toMatchSnapshot()
  })

  it('should render the NotePicker component with data', () => {
    const mockDispatch = jest.fn()
    const mockState = { currentURL: '/about' }

    const { baseElement } = render(
      WrapperProviderComp(
        mockDispatch,
        mockState
      )(<NotePicker allNotes={allNotes} isLoading={false} />)
    )
    const items = screen.getAllByText('testSubject')

    expect(items.length).toBe(3)
    expect(baseElement).toMatchSnapshot()
  })

  it('should render the NotePicker component with NO DATA', async () => {
    fetchMock.once([])
    const mockDispatch = jest.fn()
    const mockState = { currentURL: '/view-all' }
    const { baseElement } = render(
      WrapperProviderComp(
        mockDispatch,
        mockState
      )(
        <ViewAll transitionStyles={transitionStyles}>
          <NotePicker allNotes={[]} isLoading={false} />
        </ViewAll>
      )
    )

    const element = screen.getByText('All Notes')
    const message = await screen.findByText('No notes to show')

    expect(element).toBeInTheDocument()

    expect(message).toBeInTheDocument()
  })

  it('should render the About component', () => {
    const mockDispatch = jest.fn()
    const mockState = { currentURL: '/about' }
    const { baseElement } = render(
      WrapperProviderComp(
        mockDispatch,
        mockState
      )(<About transitionStyles={transitionStyles} />)
    )

    const element = screen.getByText('--About the solution --')

    expect(element).toBeInTheDocument()
    expect(baseElement).toMatchSnapshot()
  })

  it('should render the Find component', () => {
    const mockDispatch = jest.fn()
    const mockState = { currentURL: '/find' }
    const { baseElement } = render(
      WrapperProviderComp(
        mockDispatch,
        mockState
      )(<Find transitionStyles={transitionStyles} />)
    )

    const element = screen.getByText('Find Section Here')

    expect(element).toBeInTheDocument()
    expect(baseElement).toMatchSnapshot()
  })
})
