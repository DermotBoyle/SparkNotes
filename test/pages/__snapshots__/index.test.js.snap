// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainComponent/> should render the About component 1`] = `
<body>
  <div>
    <section>
      <p>
        This is an about App I built with NextJS and MongoDB
      </p>
    </section>
  </div>
</body>
`;

exports[`<MainComponent/> should render the Find component 1`] = `
<body>
  <div>
    <section>
      <p>
        Find Section Here
      </p>
    </section>
  </div>
</body>
`;

exports[`<MainComponent/> should render the side menu 1`] = `
<body>
  <div>
    <div
      class="container"
      data-testid="side-menu"
    >
      <main
        class="control-container"
      >
        <div
          class="main-view"
        >
          <section
            class="home"
            style="transition: opacity 800ms ease-out; opacity: 1;"
          >
            <h3>
              Welcome to the note App !
            </h3>
            <p>
              Please select an action from the menu
            </p>
          </section>
        </div>
        <nav
          class="side-menu"
        >
          <ul
            class="nav-list"
          >
            <li
              class="nav-link__active"
              data-testid="home"
            >
              <a
                href="/home"
              >
                home
              </a>
            </li>
            <li
              class="nav-link"
              data-testid="view all"
            >
              <a
                href="/home?section=view-all"
              >
                view all
              </a>
            </li>
            <li
              class="nav-link"
              data-testid="find"
            >
              <a
                href="/home?section=find"
              >
                find
              </a>
            </li>
            <li
              class="nav-link"
              data-testid="about"
            >
              <a
                href="/home?section=about"
              >
                about
              </a>
            </li>
          </ul>
        </nav>
        <button
          aria-label="create new note"
          class="create-button"
          data-for="create-button"
          data-tip="true"
        >
          <div
            style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
          >
            <div
              style="box-sizing: border-box; display: block; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                role="presentation"
                src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNzIiIGhlaWdodD0iMTYiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
              />
            </div>
            <noscript />
            <img
              alt="plus symbol"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
          </div>
        </button>
      </main>
      <footer
        class="footer"
      >
        <a
          href="https://github.com/dermotboyle"
          rel="noopener noreferrer"
          target="_blank"
        >
          Designed and Developed by Dermot Paul Boyle
        </a>
        <div
          class="deployment"
        >
          <p>
            Deployed on 
          </p>
          <div
            style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
          >
            <div
              style="box-sizing: border-box; display: block; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                role="presentation"
                src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNzIiIGhlaWdodD0iMTYiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
              />
            </div>
            <noscript />
            <img
              alt="Vercel Logo"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
          </div>
        </div>
      </footer>
    </div>
  </div>
</body>
`;

exports[`<MainComponent/> should render the view-all component with NO DATA 1`] = `
<body>
  <div>
    <section
      class="view-all"
    >
      <p>
        All Notes
      </p>
      <div
        class="carousel"
      >
        <h3>
          No notes to show
        </h3>
      </div>
    </section>
  </div>
</body>
`;
